src/main/java/io/confluent/connect/elasticsearch/OffsetTracker.java

```diff
@@ -0,0 +1,151 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Tracks processed records to calculate safe offsets to commit.
+ * <br>
+ * Since ElasticsearchClient can potentially process multiple batches asynchronously for the same
+ * partition, if we don't want to wait for all in-flight batches at the end of the put call
+ * (or flush/preCommit) we need to keep track of what's the highest offset that is safe to commit.
+ * For now, we do that at the individual record level because batching is handled by BulkProcessor,
+ * and we don't have control over grouping/ordering.
+ */
+class OffsetTracker {
+
+  private static final Logger log = LoggerFactory.getLogger(OffsetTracker.class);
+
+  private final Map<TopicPartition, Map<Long, OffsetState>> offsetsByPartition = new HashMap<>();
+  private final Map<TopicPartition, Long> maxOffsetByPartition = new HashMap<>();
+
+  private final AtomicLong numEntries = new AtomicLong();
+
+  static class OffsetState {
+
+    private final long offset;
+    private volatile boolean processed;
+
+    OffsetState(long offset) {
+      this.offset = offset;
+    }
+
+    /**
+     * Marks the offset as processed (ready to report to preCommit)
+     */
+    public void markProcessed() {
+      processed = true;
+    }
+
+    public boolean isProcessed() {
+      return processed;
+    }
+  }
+
+  /**
+   * Partitions are no longer owned, we should release all related resources.
+   */
+  public synchronized void closePartitions(Collection<TopicPartition> topicPartitions) {
+    topicPartitions.forEach(tp -> {
+      Map<Long, OffsetState> offsets = offsetsByPartition.remove(tp);
+      if (offsets != null) {
+        numEntries.getAndAdd(offsets.size());
+      }
+      maxOffsetByPartition.remove(tp);
+    });
+  }
+
+  /**
+   * This method assumes that new records are added in offset order.
+   * Older records can be re-added, and the same Offset object will be return if its
+   * offset hasn't been reported yet.
+   */
+  public synchronized OffsetState addPendingRecord(SinkRecord sinkRecord) {
+    log.trace("Adding pending record");
+    TopicPartition tp = new TopicPartition(sinkRecord.topic(), sinkRecord.kafkaPartition());
+    Long partitionMax = maxOffsetByPartition.get(tp);
+    if (partitionMax == null || sinkRecord.kafkaOffset() > partitionMax) {
+      numEntries.incrementAndGet();
+      return offsetsByPartition
+              // Insertion order needs to be maintained
+              .computeIfAbsent(tp, key -> new LinkedHashMap<>())
+              .computeIfAbsent(sinkRecord.kafkaOffset(), OffsetState::new);
+    } else {
+      return new OffsetState(sinkRecord.kafkaOffset());
+    }
+  }
+
+  /**
+   * @return overall number of entries
+   */
+  public long numEntries() {
```

Comment:
Can this be package protected since it's only used in tests?

Also, what is meant by "number of entries"? The total number of entries that have be processed? The number of offsets that are in `offsetsByPartition`?

Since this is only used by tests, is it worth it to maintain a separate field, especially if that field is ambiguous in its purpose? Plus, if there is a bug in computing it, then this additional complexity is definitely not worth it. Why not just make the method synchronized (like all of the others) and compute the total number of offsets?

Reply:
Renamed and updated the javadoc